source/MaterialXShaderGen/ShaderGenerators/Osl/OslSyntax.cpp

```diff
@@ -1,12 +1,124 @@
+            }
+        }
+
+        string OslStructTypeSyntax::getValue(const vector<string>& values, bool uniform) const override
+        {
+            if (values.empty())
+            {
+                throw ExceptionShaderGenError("No values given to construct a value");
+            }
+
+            string result = uniform ? "{" : getName() + "(" + values[0];
+            for (size_t i = 1; i<values.size(); ++i)
+            {
+                result += ", " + values[i];
+            }
+            result += uniform ? "}" : ")";
+
+            return result;
+        }
+    };
+
+    // For the color4 type we need even more specialization since it's a struct of a struct:
+    //
+    // struct color4 {
+    //    color rgb;
+    //    float a;
+    // }
+    //
+    class OslColor4TypeSyntax : public OslStructTypeSyntax
+    {
+    public:
+        OslColor4TypeSyntax() 
+            : OslStructTypeSyntax("color4", "color4(color(0.0), 0.0)", "{color(0.0), 0.0}", EMPTY_STRING, OslSyntax::COLOR4_MEMBERS)
+        {}
+
+        string getValue(const Value& value, bool uniform) const override
+        {
+            std::stringstream ss;
+
+            // Set float format and precision for the stream
+            const Value::FloatFormat fmt = Value::getFloatFormat();
+            ss.setf(std::ios_base::fmtflags(
+                (fmt == Value::FloatFormatFixed ? std::ios_base::fixed :
+                (fmt == Value::FloatFormatScientific ? std::ios_base::scientific : 0))),
+                std::ios_base::floatfield);
+            ss.precision(Value::getFloatPrecision());
+
+            const Color4 c = value.asA<Color4>();
+
+            if (uniform)
+            {
+                ss << "{color(" << c[0] << ", " << c[1] << ", " << c[2] << "), " << c[3] << "}";
+            }
+            else
+            {
+                ss << "color4(color(" << c[0] << ", " << c[1] << ", " << c[2] << "), " << c[3] << ")";
+            }
+
+            return ss.str();
+        }
+
+        string getValue(const vector<string>& values, bool uniform) const override
+        {
+            if (values.size() < 4)
+            {
+                throw ExceptionShaderGenError("Too few values given to construct a color4 value");
+            }
+
+            if (uniform)
+            {
+                return "{color(" + values[0] + ", " + values[1] + ", " + values[2] + "), " + values[3] + "}";
+            }
+            else
+            {
+                return "color4(color(" + values[0] + ", " + values[1] + ", " + values[2] + "), " + values[3] + ")";
+            }
+        }
+    };
+}
+
+const string OslSyntax::OUTPUT_QUALIFIER = "output";
+const vector<string> OslSyntax::VECTOR_MEMBERS  = { "[0]", "[1]", "[2]" };
+const vector<string> OslSyntax::VECTOR2_MEMBERS = { ".x", ".y" };
+const vector<string> OslSyntax::VECTOR4_MEMBERS = { ".x", ".y", ".z", ".w" };
+const vector<string> OslSyntax::COLOR2_MEMBERS  = { ".r", ".a" };
```

Comment:
Guess this was what you were mentioning that we are stuck with different syntax for color vs vec.

Reply:
Yes, the whole syntax handling of types and swizzling of values would have been orders of magnitude simpler if OSL could just have 2- and 4- component vectors and colors as built in types, like any other shading language :(